### PR TITLE
Script to update new cluster OIDC reference

### PR DESCRIPTION
**Pull request by satr**

### CODE REVIEW

File: scripts/aks/create-federated-credentials.sh [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+          echo "  - application: $displayName, clientId: $clientId, namespace:${namespace}, component:${componentName}"
+
+          fedCreds=$(az ad app federated-credential list --id "$clientId" -o json|jq -r ".[]|select(.issuer|contains(\"azure.com/$tenant/\"))|select(.subject==\"system:serviceaccount:${namespace}:${componentName}-sa\")"|jq -s '.')
+          if [[ -z "$fedCreds" ]]; then
+            echo "fail to get federated credentials"
+            exit 1
+          elif [ $(echo "$fedCreds"|jq '. | length') -eq 0 ]; then
+            echo "    - no federated credential(s) found"
+            continue
+          fi
```
Comment (nilsgstrabo) [inline scripts/aks/create-federated-credentials.sh:110]: Don't think it is neccessary to do this check. The script should register the federated credentials even if none already exist.

File: scripts/aks/create-federated-credentials.sh [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+          fedCredsCount=$(echo "$fedCreds"|jq -s ".|length")
+          echo "    - found $fedCredsCount federated credential(s)"
+
+          requiredFedCredCount=$(echo "$fedCreds"|jq -r ".[]|select(.issuer|contains(\"$ISSUER_URL\"))"|jq -s ".|length")
```
Comment (nilsgstrabo) [inline scripts/aks/create-federated-credentials.sh:115]: You should check existiance of both issuer URL and subject here.

File: scripts/aks/create-federated-credentials.sh [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            exit 1
+          elif [ $(echo "$fedCreds"|jq '. | length') -eq 0 ]; then
+            echo "    - no federated credentials found"
+            continue
+          fi
+
+          fedCredsCount=$(echo "$fedCreds"|jq -s ".|length")
+          echo "    - found $fedCredsCount federated credential(s)"
+
+          requiredFedCredCount=$(echo "$fedCreds"|jq -r ".[]|select(.issuer|contains(\"$ISSUER_URL\"))"|jq -s ".|length")
```
Comment (nilsgstrabo) [inline scripts/aks/create-federated-credentials.sh:151]: Need to check both issuer url and subject (One managed identity can be used by multiple components)

**Review by nilsgstrabo — CHANGES_REQUESTED** [review]
Added some comments.
I also think it would be useful to have a dry-run/verify-only argument so we can get a list of apps/envs/components+jobs and if the clientId is correctly configured

File: scripts/aks/create-federated-credentials.sh [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+          fedCreds=$(az identity federated-credential list --identity-name "$displayName" --resource-group "$resourceGroup")
+
+          if [[ -z "$fedCreds" ]]; then
+            echo "    - fail to get federated credential(s)"
+            exit 1
+          elif [ $(echo "$fedCreds"|jq '. | length') -eq 0 ]; then
+            echo "    - no federated credentials found"
+            continue
+          fi
```
[the remark this answers is not in the document]
Comment (emirgens) [reply scripts/aks/create-federated-credentials.sh:146]: Agree

File: scripts/aks/create-federated-credentials.sh (hunk 1 above)
Comment (emirgens) [reply scripts/aks/create-federated-credentials.sh:110]: Agree

File: scripts/aks/create-federated-credentials.sh [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+        sps=$(az ad sp list --filter "appId eq '$clientId'" --query "[].{appId:appId,displayName:displayName,objectId:id,type:servicePrincipalType}"|jq -rc '.|select(length > 0)|.[0]')
+        if [[ -z "$sps" ]]; then
+          echo "  - not found service principal for the clientId $clientId"
+          continue
+        fi
+
+        displayName=$(echo "$sps"|jq -r .displayName)
+        newFedCredName="${namespace}-${componentName}-${ISSUER_URL:(-37):36}"
+        newFedCredSubject="system:serviceaccount:${namespace}:${componentName}-sa"
```
[the remark this answers is not in the document]
Comment (emirgens) [reply scripts/aks/create-federated-credentials.sh:97]: Keep it simple

[the remark this answers is not in the document]
Comment (satr) [reply scripts/aks/create-federated-credentials.sh]: Fixed

File: scripts/aks/create-federated-credentials.sh (hunk 2 above)
Comment (satr) [reply scripts/aks/create-federated-credentials.sh]: Fixed (this code was deleted now)

File: scripts/aks/create-federated-credentials.sh (hunk 3 above)
Comment (satr) [reply scripts/aks/create-federated-credentials.sh]: Fixed (this code was deleted now)

File: scripts/aks/create-federated-credentials.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (satr) [reply scripts/aks/create-federated-credentials.sh]: Fixed

File: scripts/aks/create-federated-credentials.sh [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+  fi
+
+  echo "found environment(s) for an application: $appName"
+  for envName in $envNames
+    do
+      identityPropList=($(curl -X GET \
+          "${apiUrl}/applications/${appName}/environments/${envName}" \
+          -H "Authorization: Bearer ${token}" \
+          -H "Content-Type: application/json" 2>/dev/null | \
+          jq -r '.|select(.activeDeployment != null)|.activeDeployment as $depl|select($depl.components != null)|$depl.components[] as $comp|select($comp.identity != null)|select($comp.identity.azure != null)|{namespace:$depl.namespace,componentName:$comp.name,clientId:$comp.identity.azure.clientId}'|jq -c '.'))
```
[the remark this answers is not in the document]
Comment (satr) [reply scripts/aks/create-federated-credentials.sh]: Jobs are in the same `components` list, as regular components, so they are in use

**Review by nilsgstrabo — APPROVED** [review]

**Pull request merged**